File: app/src/main/java/com/asdamp/adapters/DateListAdapter.java
package com.asdamp.adapters;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.core.graphics.ColorUtils;
import androidx.palette.graphics.Palette;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;
import android.text.style.RelativeSizeSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.asdamp.exception.DateNotFoundException;
import com.asdamp.utility.UserInfoUtility;
import com.asdamp.x_day.Costanti;
import com.asdamp.x_day.Data;
import com.asdamp.x_day.GlideApp;
import com.asdamp.x_day.R;
import com.jaychang.st.SimpleText;
import com.omega_r.libs.omegarecyclerview.OmegaRecyclerView;
import com.omega_r.libs.omegarecyclerview.sticky_decoration.StickyAdapter;
import com.pixplicity.easyprefs.library.Prefs;

import org.threeten.bp.Month;
import org.threeten.bp.format.TextStyle;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.hdodenhof.circleimageview.CircleImageView;

import static org.apache.commons.lang3.StringUtils.capitalize;

public class DateListAdapter extends RecyclerView.Adapter<DateListAdapter.ViewHolder>  implements StickyAdapter<DateListAdapter.HeaderHolder>

{
    private ArrayList<Data> date;
    private OnListItemClickListener callback;

    @Override
    public long getStickyId(int position) {
        if(Prefs.getString("sortby","time").equalsIgnoreCase("time")){
            if(position>0)
                return Long.parseLong(date.get(position).getMonth()+""+date.get(position).getYear());
            return Long.parseLong(date.get(0).getMonth()+""+date.get(0).getYear());
        }
        else if(Prefs.getString("sortby","time").equalsIgnoreCase("alphabetical")){
            if(position>0)
                if(date.get(position).getDescrizioneIfExists()!=null && !date.get(position).getDescrizioneIfExists().equals(""))
                    return date.get(position).getDescrizioneIfExists().charAt(0);
                else return '#';
            else{
                    if(date.get(0).getDescrizioneIfExists()!=null && !date.get(0).getDescrizioneIfExists().equals(""))
                        return date.get(0).getDescrizioneIfExists().charAt(0);
                    else return '#';

            }
        }
        else if(Prefs.getString("sortby","time").equalsIgnoreCase("color")){
            if(position>0)
                return date.get(position).getColor();
            return date.get(0).getColor();
        }

        return '#';
    }

    @Override
    public HeaderHolder onCreateStickyViewHolder(ViewGroup parent) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.sticky_header_test, parent, false);
        return new HeaderHolder(view);
    }

    @Override
    public void onBindStickyViewHolder(HeaderHolder viewHolder, int position) {
        Context context=viewHolder.itemView.getContext();

        if(Prefs.getString("sortby","time").equalsIgnoreCase("time")){
            String monthName = Month.of(date.get(position).getMonth()+1).getDisplayName(TextStyle.FULL_STANDALONE, context.getResources().getConfiguration().locale) + " " + date.get(position).getYear();
            monthName=capitalize(monthName);
            viewHolder.text.setText(monthName);
            viewHolder.text.setVisibility(View.VISIBLE);

            viewHolder.color.setVisibility(View.GONE);
        }
        else if(Prefs.getString("sortby","time").equalsIgnoreCase("alphabetical")){
            viewHolder.text.setText(""+(char) getStickyId(position));
            viewHolder.color.setVisibility(View.GONE);
            viewHolder.text.setVisibility(View.VISIBLE);


        }
        else if(Prefs.getString("sortby","time").equalsIgnoreCase("color")){
            viewHolder.color.setVisibility(View.VISIBLE);
            viewHolder.text.setVisibility(View.GONE);

            viewHolder.color.setImageDrawable(new ColorDrawable((int) getStickyId(position)));

        }



    }

    public interface OnListItemClickListener{
        void onListItemClick(View v,int i);
        boolean onListItemLongClick(View v,int i);
    }
    static class ViewHolder extends RecyclerView.ViewHolder {
        // each data item is just a string in this case
        @BindView(R.id.data)
        TextView mDate;
        @BindView(R.id.mancante)
        TextView mLeft;
        @BindView(R.id.descrizionePersonale)
        TextView mDescription;
        @BindView(R.id.iv_date_image)
        ImageView mImage;


        ViewHolder(View v) {
            super(v);
            ButterKnife.bind(this, v);
            //root=v;
        }
    }
    public DateListAdapter(ArrayList<Data> arraylist)
    {
        date = arraylist;
    }

    public void setOnListItemClickListener(OnListItemClickListener callback){
        this.callback=callback;
    }
    // Create new views (invoked by the layout manager)
    @NonNull
    @Override
    public DateListAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent,
                                                         int viewType) {
        // create a new view
        View v = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.principale, parent, false);
        return new ViewHolder(v);
    }
    // Replace the contents of a view (invoked by the layout manager)
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int i) {
        Context context=holder.itemView.getContext();
        Data data = date.get(i);
        CardView cardView= (CardView) holder.itemView;
        cardView.setCardBackgroundColor(data.getColor());

        Palette p = null;
        holder.mImage.setImageDrawable(null);
        if(data.getImage()!=null)
            GlideApp.with(context).load(data.getImage()).centerCrop().into(holder.mImage);
      /*  else
            holder.mImage.setVisibility(View.INVISIBLE
            );*/

        holder.mDate.setText(date.get(i).toString());
        try{
            String lefttext=data.aggiorna(context);

            holder.mLeft.setText(UserInfoUtility.makeSpannable(lefttext, "\\d+"));
        }
        catch (ArithmeticException e){
            holder.mLeft.setText(context.getResources().getQuantityString(R.plurals.Secondi, Integer.MAX_VALUE)+"+");
        }
        String s = data.getDescrizioneIfExists();
        if(s.equalsIgnoreCase(""))
        {
            holder.mDescription.setVisibility(View.GONE);
    } else
        {
            holder.mDescription.setVisibility(View.VISIBLE);
            holder.mDescription.setText(s);
        }
        if(callback!=null) {
            holder.itemView.setOnClickListener(v -> callback.onListItemClick(holder.itemView,i));
            holder.itemView.setOnLongClickListener(v -> callback.onListItemLongClick(holder.itemView,i));

        }

    }
    public synchronized void sortList(Comparator<Data> comparator, boolean reverse){
        Collections.sort(date, comparator);
        if(reverse)
            Collections.reverse(date);

    }

    @Override
    public int getItemCount() {
        return date.size();
    }


    static class HeaderHolder extends OmegaRecyclerView.ViewHolder {
        public TextView text;
        public CircleImageView color;

        public HeaderHolder(View itemView) {
            super(itemView);
            text = (TextView) itemView.findViewById(R.id.list_view_section_header);
            color= itemView.findViewById(R.id.list_view_section_color);
        }
    }
}

File: app/src/main/java/com/asdamp/adapters/DateListAdapter.kt
package com.asdamp.adapters

import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.palette.graphics.Palette
import androidx.recyclerview.widget.RecyclerView
import com.asdamp.adapters.DateListAdapter.HeaderHolder
import com.asdamp.utility.UserInfoUtility
import com.asdamp.x_day.Data
import com.asdamp.x_day.GlideApp
import com.asdamp.x_day.R
import com.omega_r.libs.omegarecyclerview.OmegaRecyclerView
import com.omega_r.libs.omegarecyclerview.sticky_decoration.StickyAdapter
import com.pixplicity.easyprefs.library.Prefs
import de.hdodenhof.circleimageview.CircleImageView
import kotlinx.android.synthetic.main.principale.view.*
import kotlinx.android.synthetic.main.sticky_header_test.view.*
import org.apache.commons.lang3.StringUtils
import org.threeten.bp.Month
import org.threeten.bp.format.TextStyle
import java.util.*

class DateListAdapter(private val date: MutableList<Data>) : RecyclerView.Adapter<DateListAdapter.ViewHolder>(), StickyAdapter<HeaderHolder> {
    private var callback: OnListItemClickListener? = null
    override fun getStickyId(position: Int): Long {
        val sortMode = Prefs.getString("sortby", "time")
        when {
            sortMode.equals("time", ignoreCase = true) -> {
                return if (position > 0) (date[position].month.toString() + "" + date[position].year).toLong() else (date[0].month.toString() + "" + date[0].year).toLong()
            }
            sortMode.equals("alphabetical", ignoreCase = true) -> {
                return if (position > 0)
                    if (date[position].descrizioneIfExists != null && date[position].descrizioneIfExists != "")
                        date[position].descrizioneIfExists[0].toLong()
                    else '#'.toLong()
                else {
                    if (date[0].descrizioneIfExists != null && date[0].descrizioneIfExists != "")
                        date[0].descrizioneIfExists[0].toLong()
                    else '#'.toLong()
                }
            }
            sortMode.equals("color", ignoreCase = true) -> {
                return if (position > 0) date[position].color.toLong() else date[0].color.toLong()
            }
            else -> return '#'.toLong()
        }
    }

    override fun onCreateStickyViewHolder(parent: ViewGroup): HeaderHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.sticky_header_test, parent, false)
        return HeaderHolder(view)
    }

    override fun onBindStickyViewHolder(viewHolder: HeaderHolder, position: Int) {
        val context = viewHolder.itemView.context
        when {
            Prefs.getString("sortby", "time").equals("time", ignoreCase = true) -> {
                val currLocale = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                    context.resources.configuration.locales[0]
                } else {
                    context.resources.configuration.locale
                }
                var monthName: String? = Month.of(date[position].month + 1).getDisplayName(TextStyle.FULL_STANDALONE, currLocale) + " " + date[position].year
                monthName = StringUtils.capitalize(monthName)
                viewHolder.text.text = monthName
                viewHolder.text.visibility = View.VISIBLE
                viewHolder.color.visibility = View.GONE
            }
            Prefs.getString("sortby", "time").equals("alphabetical", ignoreCase = true) -> {
                viewHolder.text.text = getStickyId(position).toChar().toString()
                viewHolder.color.visibility = View.GONE
                viewHolder.text.visibility = View.VISIBLE
            }
            Prefs.getString("sortby", "time").equals("color", ignoreCase = true) -> {
                viewHolder.color.visibility = View.VISIBLE
                viewHolder.text.visibility = View.GONE
                viewHolder.color.setImageDrawable(ColorDrawable(getStickyId(position).toInt()))
            }
        }
    }

    interface OnListItemClickListener {
        fun onListItemClick(v: View, i: Int)
        fun onListItemLongClick(v: View, i: Int): Boolean
    }

    class ViewHolder(v: View) : RecyclerView.ViewHolder(v) {
        // each data item is just a string in this case
        var mDate: TextView = v.data
        var mLeft: TextView = v.mancante
        var mDescription: TextView = v.descrizionePersonale
        var mImage: ImageView = v.iv_date_image

    }

    fun setOnListItemClickListener(callback: OnListItemClickListener?) {
        this.callback = callback
    }

    // Create new views (invoked by the layout manager)
    override fun onCreateViewHolder(parent: ViewGroup,
                                    viewType: Int): ViewHolder {
        // create a new view
        val v = LayoutInflater.from(parent.context)
                .inflate(R.layout.principale, parent, false)
        return ViewHolder(v)
    }

    // Replace the contents of a view (invoked by the layout manager)
    override fun onBindViewHolder(holder: ViewHolder, i: Int) {
        val context = holder.itemView.context
        val data = date[i]
        val cardView = holder.itemView as CardView
        cardView.setCardBackgroundColor(data.color)
        val p: Palette? = null
        holder.mImage.setImageDrawable(null)
        if (data.image != null) GlideApp.with(context).load(data.image).centerCrop().into(holder.mImage)
        /*  else
            holder.mImage.setVisibility(View.INVISIBLE
            );*/holder.mDate.text = date[i].toString()
        try {
            val lefttext = data.aggiorna(context)
            holder.mLeft.text = UserInfoUtility.makeSpannable(lefttext, "\\d+")
        } catch (e: ArithmeticException) {
            holder.mLeft.text = context.resources.getQuantityString(R.plurals.Secondi, Int.MAX_VALUE) + "+"
        }
        val s = data.descrizioneIfExists
        if (s.equals("", ignoreCase = true)) {
            holder.mDescription.visibility = View.GONE
        } else {
            holder.mDescription.visibility = View.VISIBLE
            holder.mDescription.text = s
        }
        if (callback != null) {
            holder.itemView.setOnClickListener { v: View? -> callback?.onListItemClick(holder.itemView, i) }
            holder.itemView.setOnLongClickListener { v: View? -> callback?.onListItemLongClick(holder.itemView, i) ?: false }
        }
    }

    @Synchronized
    fun sortList(comparator: (Data, Data) -> Int, reverse: Boolean) {
        Collections.sort(date, comparator)
        if (reverse) date.reverse()
    }

    override fun getItemCount(): Int {
        return date.size
    }

    class HeaderHolder(itemView: View) : OmegaRecyclerView.ViewHolder(itemView) {
        var text: TextView = itemView.list_view_section_header
        var color: CircleImageView = itemView.list_view_section_color

    }

}